openshift/default-catalog-consistency/pkg/check/image.go

```diff
@@ -0,0 +1,111 @@
+package check
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/containers/image/v5/manifest"
+
+	specsgov1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+)
+
+// AllImageChecks returns a list of image checks to be performed on the image.
+func AllImageChecks() []ImageCheck {
+	return []ImageCheck{
+		ImageIsValidManifest(),
+		ImageHasExpectedLabelsWith(map[string]string{
+			"operators.operatorframework.io.index.configs.v1": "/configs",
+		}),
+	}
+}
+
+// ImageIsValidManifest checks if the image is a valid manifest.
+func ImageIsValidManifest() ImageCheck {
+	return ImageCheck{
+		Name: "ImageIsValidManifest",
+		Fn: func(ctx context.Context, root specsgov1.Descriptor, target oras.ReadOnlyTarget) error {
+			manifestReader, err := target.Fetch(ctx, root)
+			if err != nil {
+				return err
+			}
+			defer manifestReader.Close()
+			var imgManifest specsgov1.Manifest
+			if err := json.NewDecoder(manifestReader).Decode(&imgManifest); err != nil {
+				return err
+			}
+			switch imgManifest.MediaType {
+			case specsgov1.MediaTypeImageManifest, manifest.DockerV2Schema2MediaType:
+			default:
+				return fmt.Errorf("unrecognized manifest type: %s", imgManifest.MediaType)
+			}
+			return nil
+		},
+	}
+}
+
+// ImageHasExpectedLabelsWith checks if the image has the expected labels.
+func ImageHasExpectedLabelsWith(expectedLabels map[string]string) ImageCheck {
+	name := "ImageHasExpectedLabels"
+	if len(expectedLabels) > 0 {
+		var pairs []string
+		for k, v := range expectedLabels {
+			pairs = append(pairs, fmt.Sprintf("%s=%s", k, v))
+		}
+		sort.Strings(pairs)
+		name = fmt.Sprintf("%s[%s]", name, strings.Join(pairs, ","))
+	}
+
+	return ImageCheck{
+		Name: name,
+		Fn: func(ctx context.Context, root specsgov1.Descriptor, target oras.ReadOnlyTarget) error {
+			fetch, err := target.Fetch(ctx, root)
+			if err != nil {
+				return err
+			}
+			defer fetch.Close()
+			var m specsgov1.Manifest
+			if err := json.NewDecoder(fetch).Decode(&m); err != nil {
+				return err
+			}
+			switch m.MediaType {
+			case specsgov1.MediaTypeImageManifest, manifest.DockerV2Schema2MediaType:
```

Comment:
we perform this validation in ImageIsValidManifest, do we want/have to do it again here?

Reply:
and if we do, would it make sense to push the logic getting the manifest to some common function?

Reply:
Theoretically, we could call one check without the other, so I think it makes sense to keep the validation centralized and use it in both scenarios. That said, let me know if you think any further adjustments would be appropriate.